cli/src/main/java/hudson/cli/CLI.java

```diff
@@ -526,6 +641,141 @@ public boolean verify(String s, SSLSession sslSession) {
+                    }
+
+                    Integer exitStatus = channel.getExitStatus();
+                    return exitStatus;
+
+                }
+            } finally {
+                client.stop();
+            }
+        }
+    }
+
+    private static int plainHttpConnection(String url, List<String> args, CLIConnectionFactory factory) throws IOException, InterruptedException {
+        LOGGER.log(FINE, "Trying to connect to {0} via plain protocol over HTTP", url);
+        URL jenkins = new URL(url + "cli?remoting=false");
+        FullDuplexHttpStream streams = new FullDuplexHttpStream(jenkins, factory.authorization);
+        class ClientSideImpl extends PlainCLIProtocol.ClientSide {
+            boolean complete;
+            int exit = -1;
+            ClientSideImpl(InputStream is, OutputStream os) throws IOException {
+                super(is, os);
+                if (is.read() != 0) { // cf. FullDuplexHttpService
+                    throw new IOException("expected to see initial zero byte");
+                }
+            }
+            @Override
+            protected void onExit(int code) {
+                this.exit = code;
+                finished();
+            }
+            @Override
+            protected void onStdout(byte[] chunk) throws IOException {
+                System.out.write(chunk);
+            }
+            @Override
+            protected void onStderr(byte[] chunk) throws IOException {
+                System.err.write(chunk);
+            }
+            @Override
+            protected void handleClose() {
+                finished();
+            }
+            private synchronized void finished() {
+                complete = true;
+                notifyAll();
+            }
+        }
+        try (final ClientSideImpl connection = new ClientSideImpl(streams.getInputStream(), streams.getOutputStream())) {
+            for (String arg : args) {
+                connection.sendArg(arg);
+            }
+            connection.sendEncoding(Charset.defaultCharset().name());
+            connection.sendLocale(Locale.getDefault().toString());
+            connection.sendStart();
+            connection.begin();
+            final OutputStream stdin = connection.streamStdin();
+            new Thread("input reader") {
+                @Override
+                public void run() {
+                    try {
+                        int c;
+                        while ((c = System.in.read()) != -1) { // TODO use InputStream.available
+                           stdin.write(c);
+                        }
+                        connection.sendEndStdin();
+                    } catch (IOException x) {
+                        x.printStackTrace();
+                    }
+                }
+            }.start();
+            synchronized (connection) {
+                while (!connection.complete) {
+                    connection.wait();
```

Comment:
No need—if the user wishes to impose a timeout, they can do so with external tools. Anyway there is no particular limit to how long a command might run; consider

    java -jar jenkins-cli.jar -s … build -f -v really-slow-job